src/raftstore/store/fsm/apply.rs

```diff
@@ -577,25 +626,42 @@ impl ApplyDelegate {
             }
 
             let res = match entry.get_entry_type() {
-                EntryType::EntryNormal => self.handle_raft_entry_normal(apply_ctx, entry),
-                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(apply_ctx, entry),
+                EntryType::EntryNormal => self.handle_raft_entry_normal(apply_ctx, &entry),
+                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(apply_ctx, &entry),
             };
 
-            if let Some(res) = res {
-                results.push(res);
+            match res {
+                ApplyResult::None => {}
+                ApplyResult::Res(res) => results.push(res),
+                ApplyResult::WaitMergeSource(ready_to_merge) => {
+                    apply_ctx.committed_count -= drainer.len() + 1;
+                    let mut pending_entries = Vec::with_capacity(drainer.len() + 1);
+                    // Note that CommitMerge is skipped when `WaitMergeSource` is returned.
+                    // So we need to enqueue it again and execute it again when resuming.
+                    pending_entries.push(entry);
+                    pending_entries.extend(drainer);
+                    apply_ctx.finish_for(self, results);
+                    self.wait_merge_state = Some(WaitSourceMergeState {
```

Comment:
one is `MergeSource`, the other is `SourceMerge`. We should make it consistent.

Reply:
You read it the wrong way. It's a state about merge that wait for source.